bundles/org.openhab.binding.solarwatt/src/main/java/org/openhab/binding/solarwatt/internal/handler/EnergyManagerHandler.java

```diff
@@ -0,0 +1,408 @@
+     */
+    private final Runnable refreshRunnable = () -> {
+        EnergyManagerHandler.this.updateChannels();
+        EnergyManagerHandler.this.updateAllChildThings();
+    };
+
+    /**
+     * Create the handler.
+     *
+     * @param thing for which the handler is responsible
+     * @param channelTypeProvider provider for the channels
+     * @param httpClient connect to energy manager via this client
+     */
+    public EnergyManagerHandler(final Bridge thing, final SolarwattChannelTypeProvider channelTypeProvider,
+            final HttpClient httpClient) {
+        super(thing);
+        this.logger.debug("{} created", this);
+        this.connector = new EnergyManagerConnector(httpClient);
+        this.channelTypeProvider = channelTypeProvider;
+    }
+
+    /**
+     * Get services which are provided by this handler.
+     *
+     * Only service discovery is provided by
+     * 
+     * @return collection containing our discovery service
+     */
+    @Override
+    public Collection<Class<? extends ThingHandlerService>> getServices() {
+        return Collections.singleton(SolarwattDevicesDiscoveryService.class);
+    }
+
+    /**
+     * Execute the desired commands.
+     *
+     * Only refresh is supported and relayed to all childs of this thing.
+     *
+     * @param channelUID for which the command is issued
+     * @param command command issued
+     */
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+        this.logger.debug("handleCommand called for channel: {}, command: {}", channelUID, command);
+
+        if (command instanceof RefreshType) {
+            this.updateChannels();
+        }
+    }
+
+    /**
+     * Dynnamically updates all known channel states of the energy manager.
+     */
+    public void updateChannels() {
+        this.logger.trace("{} updateChannels", this);
+        Map<String, Device> devices = this.getDevices();
+        if (devices != null) {
+            if (this.energyManagerGuid == null) {
+                try {
+                    this.findEnergyManagerGuid(devices);
+                } catch (SolarwattConnectionException ex) {
+                    this.logger.warn("Failed updating EnergyManager channels: {}", ex.getMessage());
+                }
+            }
+            EnergyManager energyManager = (EnergyManager) devices.get(this.energyManagerGuid);
+
+            if (energyManager != null) {
+                this.calculateUpdates(energyManager);
+
+                Map<String, String> properties = this.editProperties();
+                properties.put(THING_PROPERTIES_GUID, energyManager.getGuid());
+                this.updateProperties(properties);
+                energyManager.getStateValues().forEach((stateName, stateValue) -> {
+                    this.logger.trace("{}: {} - {}", this.getThing().getUID(), stateName, stateValue.toFullString());
+                    this.updateState(stateName, stateValue);
```

Comment:
My opinion is you should remove the logger statement. When the updateState is run, the event log will have the information logged, so you are causing a higher load on the system by logging twice.

See https://www.openhab.org/docs/developer/guidelines.html#f-logging

Reply:
Excessive logging is removed.